### DISCUSSION TITLE
How does your IR encoder actually work?

### DISCUSSION THREAD

**Discussion by toaster-robotics** [adamb314/ServoProject#18] [General]
Hello,
I saw this project on YouTube and think it's awesome. I'm trying to understand how you get the IR encoder working. I tried looking through your code but I'm not that familiar with C/C++ so I thought I'd just ask. This is how I think it works:

- the offset dark circle on your encoder wheel allows for a gradient of brightness on the IR sensor while the wheel spins. 
- Let's assume the darkest value is 0 and the brightest value is 1. 
- If we start with the IR receiver directly over the darkest region (0°) and rotate to (180°) the receiver will go from 0 gradually to 1
- if we continue spinning from 180° to 360°, the receiver will go from 1 gradually back to 0. With a mirror image of a reading vs angle curve from 0°  to 180°.
- having the second IR receiver offset by 45° allows you to know if you're on the first 180° side or the 2nd. 

I built a small prototype using your disc dimensions, enclosed it and pointed a IR led and receiver at it. However, the readings on the IR receiver don't change much. It's not sensitive to the offset dark disc's position. I thought maybe the LED was too bright and saturating the receiver so I played around with its brightness using PWM. But that just screws up the receiver readings. I think it's catching different parts of the PWM so on some readings it's low and others it's high.

**Comment by toaster-robotics** [comment]
So there's almost zero information on this type of analog, absolute optical encoder on the web. Everything I could find is incremental stuff. BUT...my internet searching skills payed off (after many hours) and I found this!:

[https://ichiro-maruta.blogspot.com/2010/02/blog-post.html](https://ichiro-maruta.blogspot.com/2010/02/blog-post.html)

I'm going to replicate the setup and do a write-up in english for anyone else that's trying to figure this stuff out.

**Comment by adamb314** [comment]
The optical encoder is very much inspired by how a resolver works, so you could find more info about how that works.

The biggest difference with the optical version is that it is very difficult to get a perfect sin cos signal from the sensors. Non-linearities in the sensors and errors in the disc result in much more complicated curves. To handle this, my implementation relies on a large set of calibration values (2048), where each set of values represent one position. The microcontroller then searches through these to find the position that best fits the current sensor values.

If you prefer python over C++ you could also have a look at the `calculatePosition` function in the [opticalEncoder.py](https://github.com/adamb314/ServoProject/blob/main/Python/ServoProjectModules/CalibrationAnalyzers/OpticalEncoder.py) file to see how it works exactly.
